### ISSUE TITLE
[ENHANCEMENT] ACF 'rectangle' parallelization

### ISSUE THREAD

**Issue by AviadAvr** [pastas/pastas#846]
Currently, `_compute_ccf_rectangle()` works by utilizing nested for-loops for comparing the difference in lags to the `bin_width`. This approach works well but is substantially slower compared to the `regular` and `gaussian` methods, and especially when compared to FFT-based methods, which is understandable due to the constraint of having irregular timeseries.

A suggested improvement would be to compute each lag as a parallel process using `parallel=True` as part of the Numba `@njit` decorator.

Instead of:
```
@njit
def _compute_ccf_rectangle(
    lags: ArrayLike,
    t_x: ArrayLike,
    x: ArrayLike,
    t_y: ArrayLike,
    y: ArrayLike,
    bin_width: float = 0.5,
) -> Tuple[ArrayLike, ArrayLike]:
    """Internal numba-optimized method to compute the ccf."""
    c = empty_like(lags)
    b = empty_like(lags)
    n = len(t_x)

    for k in range(len(lags)):
        cl = 0.0
        b_sum = 0.0
        for i in range(n):
            for j in range(n):
                d = abs(t_x[i] - t_y[j]) - lags[k]
                if abs(d) <= bin_width:
                    cl += x[i] * y[j]
                    b_sum += 1
        if b_sum == 0.0:
            c[k] = nan
            b[k] = 0.01  # Prevent division by zero error
        else:
            c[k] = cl / b_sum
            b[k] = b_sum / 2  # divide by 2 because we over count in for-loop
    return c, b
```

Replace with:
```
from numba import prange

@njit(parallel=True) # Use Numba's parallel processing
def _compute_ccf_rectangle(
    """ no changes here"""

    for k in prange(len(lags)): # Use prange for parallel loops in Numba
        cl = 0.0
        b_sum = 0.0
        for i in range(n):
            """ rest of the code, no changes here"""
    return c, b
```

This could use a bit more testing, but from some benchmarking i've noticed an improvement by an order of magnitude or so. 
*This would of course vary based on the CPU used.

**Comment by martinvonk** [comment]
Hi @AviadAvr ,

Thank you for making this issue. Can you provide a small example script where you run the benchmark and test the results? This way we can check the speedup.

We have been discussing the topic of parallelization with the Pastas development team lately. We have implemented some parallel methods in Pastastore where parallelization is built around models. Also implementing parallelization within models could cause some nasty conflicts so we have to think about how large the benefits are parallelization within models..

**Comment by AviadAvr** [comment]
Hi @martinvonk ,

Sure thing:

```
import pastas as ps
import numpy as np
import pandas as pd
import time

# generate time series data with irregular time steps.
np.random.seed(42)
dates = pd.date_range(start="2010-01-01", end="2024-12-31", freq="D")
dates = dates[~(dates.month == 2)]
obs_data = np.random.normal(1, 0.1, len(dates))
obs = pd.Series(obs_data, index=dates)

dates = pd.date_range(start="2000-01-01", end="2024-12-31", freq="D")
recharge_series = pd.Series(np.random.normal(0, 0.01, len(dates)), index=dates)


ml = ps.Model(obs)
rch = ps.StressModel(recharge_series, ps.Exponential(), name="recharge")
ml.add_stressmodel(rch)
ml.solve(report=False)

t = time.time()
acf_non_parallel = ps.stats.acf(ml.residuals())
print("time to calculate acf current way:", time.time() - t) # 5.25s

t = time.time()
acf_parallel = ps.stats.acf_par(ml.residuals())
print("time to calculate acf with parallel processing:", time.time() - t) # 0.84s
```

The only changes I made to `ps.stats.acf_par()` are:
```
from numba import prange

@njit(parallel=True) # Use Numba's parallel processing
def _compute_ccf_rectangle(
    """ no changes here"""

    for k in prange(len(lags)): # Use prange for parallel loops in Numba
            """ rest of the code, no changes here"""
    return c, b 
```

Also tested this on real data.

**Comment by AviadAvr** [comment]
> Also implementing parallelization within models could cause some nasty conflicts

By the way, @martinvonk, can I ask what you meant by this?
To my understanding, the ACF calculation shouldn't have any effect on the model itself.

**Comment by martinvonk** [comment]
If we make functions within models that can do stuff in parallel, and also have functions around models that run in parallel as well. Then stuff will probably break. So I think that is the main decision we are facing with Pastas. What to parallelize, and what not? Or just parallelize everything and if it breaks, it breaks.

**Comment by raoulcollenteur** [comment]
Closing the issue as won't fix, can reopen when we decide to make njit(parallel=True_ possible :)
